### DISCUSSION TITLE
How to fix the error:  mitmproxy's console interface requires a tty. Please run mitmproxy in an interactive shell environment.

### DISCUSSION THREAD

**Discussion by moh-hassan** [Tyrrrz/CliWrap#229] [General]
In windows 10, I try to run [mitmproxy.exe](https://github.com/mitmproxy/mitmproxy)  which is a proxy server that run in an interactive mode,
I  get the  error:
```
Error: mitmproxy's console interface requires a tty. Please run mitmproxy in an interactive shell environment.
Stream does not support reading.
```
I use the next code:
```
  public static async Task StartProxyAsync()
  {
      var stdOut = Console.OpenStandardOutput();
      var stdErr = Console.OpenStandardError();
      var stdin = Console.OpenStandardError();

      var cmd = CliWrap.Cli.Wrap("mitmproxy")
          // .WithArguments($"")
          .WithStandardOutputPipe(PipeTarget.ToStream(stdOut))
          .WithStandardErrorPipe(PipeTarget.ToStream(stdErr))
          .WithStandardInputPipe(PipeSource.FromStream(stdin))
          .WithValidation(CommandResultValidation.None);

      var task = cmd.ExecuteAsync();
      await task;
  }
  ```
Also I tried to run it from within shell "cmd" or "powershell" or even as a batch file , but get the same error.

How to set tty console interface?

**Comment by Tyrrrz** [comment]
Related to #225.

You can try wrapping it with a [`winpty`](https://github.com/rprichard/winpty):

```csharp
Cli.Wrap("winpty").WithArguments(["mitmproxy", ...]);
```

**Reply by moh-hassan** [reply]
Thanks for reply.
I have already winpty  installed by Git in the folder: "C:\Program Files\Git\usr\bin\winpty.exe"

i modified the code:
```
var winpty= "C:\Program Files\Git\usr\bin\winpty.exe";
var cmd = CliWrap.Cli.Wrap(winpty)
           .WithArguments($"mitmproxy")
		   //...rest of code as is
```
I get the error:
```
stdin is not a tty
```
and it can not run mitmproxy.

**Reply by Tyrrrz** [reply]
Hm, interesting. I have not tried this myself, so I don't know what to recommend. This is an issue I've been meaning to get to eventually with #225, but it's not trivial and I didn't have the time.

Were you able to run `mitmproxy` and read its output/error using the standard `Process` class? Is it the fact that the stdin is also redirected that's the main problem? 🤔

**Reply by moh-hassan** [reply]
I tried using the **Process** class with redirecting all standard input/output = true, but get the same error:
```
stdin is not a tty
```
The strange is that I can run the next command from within `CMD` shell using **winpty**  and mitmproxy run with all its interactive features 
```
"C:\Program Files\Git\usr\bin\winpty.exe" mitmproxy
```

It seems that the child process created by Process or CliWrap runs independently and doesn't have the same interactive terminal environment as CMD.
The **stdin** (standard input) of the child process is not connected to the user's keyboard input; instead, it's connected to the input stream provided by the  C# code.
Since **mitmproxy** expects interactive input from a terminal (tty), it detects that the input is not coming from a human user and raises the **'stdin is not a tty'** error.

**Reply by Tyrrrz** [reply]
Yeah it appears so. I don't know what to suggest. If you end up figuring it out somehow, please let me know because I'm curious myself.

**Reply by moh-hassan** [reply]
I fixed the error using the ```Process``` class without the need of ```winPty``` and could apply interactive I/O  with mitmproxy and no warning of ```tty``` support.
I set   all the  redirection of standardInput, standardOutput and standardError to false and `CreateNoWindow = false`

The code is:
```
var processStartInfo = new ProcessStartInfo
{
    FileName = "mitmproxy",
    //Arguments = "-p 8080", //default port 8080
    //The next properties are false by default
    RedirectStandardInput = false, // No input redirection
    RedirectStandardOutput = false, // No output redirection
    RedirectStandardError = false, // No error redirection
    UseShellExecute = false, // Don't use shell execution
    CreateNoWindow = false // Show the process window
};

using var process = new Process
{
    StartInfo = processStartInfo
};

process.Start();
await process.WaitForExitAsync(); 
Console.WriteLine($"process.ExitCode {process.ExitCode}");        
 
```

mitmproxy is working fine with all colored screen and keyboard support (full tty support)
Tested in windows 11, net8.

I couldn't do the same using CliWrap even with setting I/O pipe to Null.

It seems that is  because CliWrap set all I/O redirection = `[true](https://github.com/Tyrrrz/CliWrap/blob/master/CliWrap/Command.Execution.cs#L80)
and [CreateNoWindow = true](https://github.com/Tyrrrz/CliWrap/blob/master/CliWrap/Command.Execution.cs#L89)

It is nice if adding this Interactive action as a new feature to the package that mimic the above code.

My suggestion: is adding a new method, e.g.  `WithInteractive() by  setting all i/o redirection = false and avoid  setting  i/o piping.

**Reply by Tyrrrz** [reply]
Yeah, without redirection it will work. But unfortunately, this means we can't read anything the program writes to the console programmatically, or provide input to it.

**Reply by moh-hassan** [reply]
That is correct. But such interactive console tools has its own output logs that can be saved to files or using scripts.
They  can be passed as arguments and these arguments can be controlled programatically per case. 

Plus, running interactive program out of box ( full tty control) with all its i/o capability using a built-in feature of Process class encourage adding this feature to CliWrap.

It is equivalant to:
```
        .WithStandardOutputPipe(PipeTarget.ToStream(Console.OpenStandardOutput()))
	.WithStandardErrorPipe(PipeTarget.ToStream(Console.OpenStandardError()))
	.WithStandardInputPipe(PipeSource.FromStream(Console.OpenStandardInput()
//But without coding.
```
This feature can resolve issues like: [run ssh #208](https://github.com/Tyrrrz/CliWrap/discussions/208) , [cmd #197](https://github.com/Tyrrrz/CliWrap/discussions/197) and similar interactive tools.

Also, the important point, this feature can  [allocate a pseudo-terminal for running command-line applications](https://github.com/Tyrrrz/CliWrap/issues/225)
